src/uct/ib/rdmacm/rdmacm_cm.c

```diff
@@ -113,6 +115,254 @@ void uct_rdmacm_cm_cqs_cleanup(uct_rdmacm_cm_t *cm)
+        }
+    }
+
+end:
+    pthread_mutex_unlock(&ctx->mutex);
+#endif
+
+    return status;
+}
+
+static void uct_rdmacm_cm_reserved_qpn_context_cleanup(uct_rdmacm_cm_t *cm)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    ucs_status_t status;
+
+    kh_foreach_value(&cm->ctxs, ctx, {
+        status = uct_rdmacm_cm_reserved_qpn_blks_free(ctx);
+        if (status != UCS_OK) {
+            ucs_error("qpn_context_cleanup failed: %m");
+        }
+        ucs_free(ctx);
+    });
+
+    kh_destroy_inplace(uct_rdmacm_cm_qpn_contexts, &cm->ctxs);
+}
+
+ucs_status_t uct_rdmacm_cm_get_reserved_qpn_context(uct_rdmacm_cm_t *cm,
+                                                    struct ibv_context *verbs,
+                                                    struct uct_rdmacm_cm_reserved_qpn_context **ctx_p)
+{
+    struct uct_rdmacm_cm_reserved_qpn_context *ctx;
+    khiter_t iter;
+    int ret;
+
+    iter = kh_put(uct_rdmacm_cm_qpn_contexts, &cm->ctxs,
+                  ibv_get_device_guid(verbs->device), &ret);
+    if (ret == -1) {
+        ucs_error("cm %p: cannot allocate hash entry for reserved qpn context", cm);
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (ret == 0) {
+        /* already exists so use it */
+        ctx = kh_value(&cm->ctxs, iter);
+    } else {
+        /* Create a reserved qpn context */
+        ctx = ucs_malloc(sizeof(struct uct_rdmacm_cm_reserved_qpn_context),
+                         "reserved_qpn_context");
+        if (ctx == NULL) {
+            kh_del(uct_rdmacm_cm_qpn_contexts, &cm->ctxs, iter);
+            ucs_error("cm %p: malloc reserved qpn context failed: %m", cm);
+            return UCS_ERR_NO_MEMORY;
+        }
+
+        pthread_mutex_init(&ctx->mutex, NULL);
+        ucs_list_head_init(&ctx->blk_list);
+        uct_rdmacm_cm_query_reserved_qpn_cap(verbs, ctx);
+        uct_rdmacm_cm_query_reserved_qpn_cap_2(verbs, ctx);
+
+        kh_value(&cm->ctxs, iter) = ctx;
+    }
+
+    *ctx_p = ctx;
+    return UCS_OK;
+}
+
+ucs_status_t
+uct_rdmacm_cm_reserved_qpn_blk_alloc(uct_rdmacm_cm_t *cm,
+                                     struct ibv_context *verbs,
+                                     struct uct_rdmacm_cm_reserved_qpn_context *ctx,
+                                     struct uct_rdmacm_cm_reserved_qpn_blk **blk_p)
+{
+    ucs_status_t status = UCS_OK;
+
+#if HAVE_DECL_MLX5DV_IS_SUPPORTED
+    char in[UCT_IB_MLX5DV_ST_SZ_BYTES(create_reserved_qpn_in)]   = {};
+    char out[UCT_IB_MLX5DV_ST_SZ_BYTES(general_obj_out_cmd_hdr)] = {};
+    struct uct_rdmacm_cm_reserved_qpn_blk *blk;
+    void *attr;
+
+    if((1 << ctx->log_reserved_qpn_granularity) > (sizeof(*blk->bmp) << 3)) {
+        ucs_error("reserved qpn bit map length is not enough");
+        return UCS_ERR_INVALID_PARAM;
+    };
+
+    blk = ucs_calloc(1, sizeof(*blk), "reserved_qpn_blk");
+    if (!blk) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    blk->bmp = ucs_malloc(sizeof(*blk->bmp), "reserved_qpn_blk_bmp");;
```

Comment:
can we make it plain structure to avoid extra allocation and memory fragmentation? seems like we need these anyway

Reply:
@evgeny-leksikov Changed bmp as plain structure since each blk anyway need one bmp. The blk number is dynamic. It depends on how many blks are in use. So it should use ucs_malloc/ucs_free.